packages/math/src/SphericalHarmonics3.ts

```diff
@@ -0,0 +1,242 @@
+
+/**
+ * Use SH3 to represent irradiance environment maps efficiently, allowing for interactive rendering of diffuse objects under distant illumination.
+ * @remarks
+ * https://graphics.stanford.edu/papers/envmap/envmap.pdf
+ * http://www.ppsloan.org/publications/StupidSH36.pdf
+ * https://google.github.io/filament/Filament.md.html#annex/sphericalharmonics
+ */
+export class SphericalHarmonics3 implements IClone {
+  static fromArray(shArray: number[], out: SphericalHarmonics3): void {
+    if (shArray.length != 27) {
+      console.error("sh coefficients must be as large as 27");
+    }
+
+    out.y00.setValue(shArray[0], shArray[1], shArray[2], 0);
+    out.y1_1.setValue(shArray[3], shArray[4], shArray[5], 0);
+    out.y10.setValue(shArray[6], shArray[7], shArray[8], 0);
+    out.y11.setValue(shArray[9], shArray[10], shArray[11], 0);
+    out.y2_2.setValue(shArray[12], shArray[13], shArray[14], 0);
+    out.y2_1.setValue(shArray[15], shArray[16], shArray[17], 0);
+    out.y20.setValue(shArray[18], shArray[19], shArray[20], 0);
+    out.y21.setValue(shArray[21], shArray[22], shArray[23], 0);
+    out.y22.setValue(shArray[24], shArray[25], shArray[26], 0);
+  }
+
+  private static _basisFunction = [
+    0.282095, //  1/2 * Math.sqrt(1 / PI)
+
+    -0.488603, // -1/2 * Math.sqrt(3 / PI)
+    0.488603, //  1/2 * Math.sqrt(3 / PI)
+    -0.488603, // -1/2 * Math.sqrt(3 / PI)
+
+    1.092548, //  1/2 * Math.sqrt(15 / PI)
+    -1.092548, // -1/2 * Math.sqrt(15 / PI)
+    0.315392, //  1/4 * Math.sqrt(5 / PI)
+    -1.092548, // -1/2 * Math.sqrt(15 / PI)
+    0.546274 //  1/4 * Math.sqrt(15 / PI)
+  ];
+
+  private static _convolutionKernel = [
+    3.141593, //  PI
+    2.094395, // (2 * PI) / 3,
+    0.785398 //   PI / 4
+  ];
+
+  private static _tempColor = new Color();
+
+  /**  The Y(0, 0) coefficient of the SH3. */
+  y00: Color = new Color(0, 0, 0, 0);
+  /**  The Y(1, -1) coefficient of the SH3. */
+  y1_1: Color = new Color(0, 0, 0, 0);
+  /**  The Y(1, 0) coefficient of the SH3. */
+  y10: Color = new Color(0, 0, 0, 0);
+  /**  The Y(1, 1) coefficient of the SH3. */
+  y11: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, -2) coefficient of the SH3. */
+  y2_2: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, -1) coefficient of the SH3. */
+  y2_1: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, 0) coefficient of the SH3. */
+  y20: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, 1) coefficient of the SH3. */
+  y21: Color = new Color(0, 0, 0, 0);
+  /**  The Y(2, 2) coefficient of the SH3. */
+  y22: Color = new Color(0, 0, 0, 0);
+
+  private _coefficients: Float32Array = new Float32Array(27);
+
+  /**
+   * Get pre-scaled coefficients used in shader.
+   * @remarks
+   * Convert radiance to irradiance with the A_l which is convoluted by the cosine lobe and pre-scale the basis function.
+   * Reference equation [4,5,6,7,8,9] from https://graphics.stanford.edu/papers/envmap/envmap.pdf
+   */
+  get preScaledCoefficients(): Float32Array {
```

Comment:
Your comments write `Convert radiance to irradiance` already, so why you named this funtion name 'preScaledCoefficients'. This is very strange.

`convertRadianceToIrradiance` is better.

Reply:
+1